File: src/leetcode/python/medium/MaximumProductOfWordLengths_318.py
class MaximumProductOfWordLengths_318(object):
    def maxProduct(self, words):
        """
        :type words: List[str]
        :rtype: int
        """
        test = {}
        for w in words:
            temp = 0x0
            for ch in w:
                temp |= (1 << ord(ch) - 97)
            test[temp] = max(test.get(temp, 0), len(w))
        return max([test[x] * test[y] for x in test for y in test if not x & y] or [0])

File: src/leetcode/java/MaximumProductOfWordLengths_318.java
package leetcode.java.medium;

public class MaximumProductOfWordLengths_318 {
  public int maxProduct(String[] words) {
    int[] wordBits = new int [words.length];
    for (int i = 0; i < wordBits.length; i++) {
      for (char w : words[i].toCharArray()) {
        wordBits[i] |= 1 << (w - 'a');
      }
    }
    
    int maxProduct = 0;
    for (int i = 0; i < words.length; i++) {
      for (int j = i + 1; j < words.length; j++) {
        if ((wordBits[i] & wordBits[j]) == 0) {
          maxProduct = Math.max(maxProduct, words[i].length() * words[j].length());
        }
      }
    }
    return maxProduct;
  }
}
